<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [2]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 

# параметры для подключения к нужной базе данных
DBNAME = 'project_sql' # название базы, к которой нужно подключиться;
USER = 'skillfactory' # имя пользователя в СУБД;
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt' # 
HOST = '84.201.134.129' # адрес, по которому нужно подключиться;
PORT = 5432 # порт, к которому нужно подключиться (по умолчанию равен 5432).

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса
query_3_1 = f'''SELECT
COUNT(id) as "Всего вакансий в базе"
FROM vacancies
'''

In [5]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_24208\1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


Всего вакансий в базе
0                  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
# текст запроса
query_3_2 = f'''SELECT
COUNT(id) as "Всего работодателей"
FROM employers
'''

In [7]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_24208\3062646138.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


Всего работодателей
0                23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# текст запроса
query_3_3 = f'''SELECT
COUNT(id) as "Всего городов с вакансиями"
FROM areas
'''

In [9]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_24208\928149064.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


Всего городов с вакансиями
0                        1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# текст запроса
query_3_4 = f'''SELECT
COUNT(id) as "Сферы деятельности работодателей"
FROM industries
'''

In [11]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_24208\2101691108.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


Сферы деятельности работодателей
0                               294

***

In [12]:
# выводы по предварительному анализу данных
total_numbers = f'''SELECT *
FROM 
(SELECT
COUNT(id) as "Вакансии"
FROM vacancies) as vacancies,
(SELECT
COUNT(id) as "Работодатели"
FROM employers) as employers,
(SELECT
COUNT(id) as "Города"
FROM areas) as areas,
(SELECT
COUNT(id) as "Сферы деятельности"
FROM industries) as Industries
'''

df = pd.read_sql_query(total_numbers, connection)
df







C:\Users\asus\AppData\Local\Temp\ipykernel_24208\3999958531.py:18: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(total_numbers, connection)


Вакансии  Работодатели  Города  Сферы деятельности
0     49197         23501    1362                 294

 В наличии имеются данные о 49147 вакансиях. Вакансии размещены в 1362 городах и представлены 294 сферами и предоставлены 23501 работодателем.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [13]:
# текст запроса
query_4_1 = f'''SELECT 
    a.name "Город",
    count(*) "Кол-во вакансий"
FROM areas a 
JOIN vacancies v ON a.id = v.area_id
GROUP BY 1
ORDER BY 2 DESC
LIMIT 20'''

In [14]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_24208\2457173557.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


Город  Кол-во вакансий
0            Москва             5333
1   Санкт-Петербург             2851
2             Минск             2112
3       Новосибирск             2006
4            Алматы             1892
5      Екатеринбург             1698
6   Нижний Новгород             1670
7            Казань             1415
8         Краснодар             1301
9            Самара             1144
10   Ростов-на-Дону             1131
11          Воронеж             1063
12          Ташкент              851
13       Красноярск              847
14       Нур-Султан              840
15        Челябинск              786
16            Пермь              771
17              Уфа              767
18            Томск              692
19             Омск              617

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [15]:
# текст запроса
query_4_2 = f'''SELECT 
    COUNT(*) as "Вакансий с данными о зарплате"
FROM vacancies
WHERE salary_from is not null or salary_to is not null --фильтр когда одна из колонок с зарплатой с не нулевым значением''' 



In [16]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df


C:\Users\asus\AppData\Local\Temp\ipykernel_24208\3981930701.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


Вакансий с данными о зарплате
0                          24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [17]:
# текст запроса
query_4_3 = f'''SELECT *
FROM -- создадим два отдельных подзапроса и выведем их как одну выборку
(SELECT 
    round(avg(salary_from)) as "Средний минимум"
FROM vacancies) as salary_from,

(SELECT 
    round(avg(salary_to)) as "Средний максимум"
FROM vacancies) as salary_to'''

In [18]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_24208\2500607024.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


Средний минимум  Средний максимум
0          71065.0          110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [19]:
# текст запроса
query_4_4 = f'''SELECT 
    schedule "График работы",
    employment "Занятость",
    count(*) "Количество"
FROM vacancies
GROUP BY 1, 2
ORDER BY 3 desc'''

In [54]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
display(df.head(5))


C:\Users\asus\AppData\Local\Temp\ipykernel_24208\2456702289.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



График работы            Занятость  Количество
0       Полный день     Полная занятость       35367
1  Удаленная работа     Полная занятость        7802
2     Гибкий график     Полная занятость        1593
3  Удаленная работа  Частичная занятость        1312
4    Сменный график     Полная занятость         940

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [55]:
# текст запроса
query_4_5 = f'''SELECT 
    experience "Опыт работы",
    count(*) "Количество"
FROM vacancies
GROUP BY 1 
ORDER BY 2'''

In [56]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df
import plotly.express as px
fig = px.pie(df, values='Количество', names='Опыт работы', title='Визуализация требуемого опыта')
fig.show()

C:\Users\asus\AppData\Local\Temp\ipykernel_24208\3347421622.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



***

In [57]:
# выводы по детальному анализу вакансий
# Узнаем сколько вакакансий приходится на города миллионики

# код для получения списка городов-милионников
import pandas as pd
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
df = pd.read_html(url)[1] #в квадратных скобках индекс таблицы на странице - подбирается путем проб и ошибок
cities = tuple(df['Город']) #SQL принимает кортеж. Выводим данные колонки с городами в виде кортежа

query_4_6 = f'''(SELECT 
    a.name "Город",
    count(*) "Кол-во вакансий"
FROM areas a 
JOIN vacancies v ON a.id = v.area_id
WHERE a.name in {cities}
GROUP BY 1
ORDER BY 2 DESC)
'''

million_cities = pd.read_sql_query(query_4_6, connection) #таблица городов-миллиоников с вакансиями
#визуализация
import plotly.express as px
fig = px.pie(million_cities, values='Кол-во вакансий', names='Город', title='Количество вакансий в городах-миллиониках России')
fig.show()

#Посчитаем соклько вакансий приходится на города миллионики России и вместе с городами, которые входят в ТОП-20 по кол-ву вакансий
query_4_7 = f'''SELECT 
    *
FROM
(SELECT 
    COUNT(*) "Итого в городах-миллиониках"
FROM employers e
JOIN vacancies v ON v.employer_id = e.id 
JOIN areas a ON v.area_id = a.id
WHERE a.name in {cities}) as total,

(SELECT 
    COUNT(*) "Всего вакансий"
FROM employers e
JOIN vacancies v ON v.employer_id = e.id 
JOIN areas a ON v.area_id = a.id) as total_total'''
total = pd.read_sql_query(query_4_7, connection)
display(total)





C:\Users\asus\AppData\Local\Temp\ipykernel_24208\47022352.py:20: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



C:\Users\asus\AppData\Local\Temp\ipykernel_24208\47022352.py:42: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Итого в городах-миллиониках  Всего вакансий
0                        23856           49197

1. Пятерку городов лидеров по количеству вакансий составляют:
    * Москва
    * Питер
    * Минск
    * Новосибирск
    * Алматы
   Нужно отметить, что больше половины вакансий приходится на города-миллионики. При этом, по количеству населения вакансии распределены не равномерно. Другими словами, большое кол-во вакансий в городе не означает, что предложений от работодателя в этом городе больше в расчете на душу населения. Например, в Москве 1 вакансия приходится почти на 2,5 тысячи человек, а в Новосибирске 1 вакансия приходится на 815 человек. 
2.  Почти половина вакансий из представленной выборки имеет информацию о предоставляемой зарплате. 
3. * 71065 - средний показатель нижнего порога зарплаты 
   * 110537 - средний показатель верхнего порога зарплаты
4. * Большинство работодателей предлагают работу с полной занятостью на полный день. 
   * Приблизительно 5 часть из всех вакансий предполагают удаленную работу. 
   * Не значительная часть вакансий предлагают работу с гибким графиком, либо частичной занятостью.
5. * 53% из всех вакансий подойдет для начинающего специалиста с опытом работы от 1 до 3 лет, 
   * 30% вакансий предлагает работу для опытных специалистов с опытом от 3 лет,
   * 14.6% вакансий для соискателей без опыта работы


# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [58]:
# текст запроса
query_5_1 = f'''SELECT 
    e.name,
    COUNT(*)
FROM vacancies v 
JOIN employers e ON v.employer_id = e.id
GROUP BY 1 
ORDER BY 2 DESC
LIMIT 5 
'''

In [59]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_24208\3834399874.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  count
0         Яндекс   1933
1     Ростелеком    491
2       Тинькофф    444
3           СБЕР    428
4  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [60]:
# текст запроса
query_5_2 = f'''SELECT 
    a.name as "Название города",
    COUNT(distinct e.id) as "кол-во работодателей",
    COUNT(distinct v.id) as "кол-во вакансий"
FROM areas a 
LEFT JOIN employers e ON a.id = e.area 
LEFT JOIN vacancies v ON v.area_id = a.id
WHERE v.id is null
GROUP BY 1
ORDER BY 2 DESC
LIMIT 10
'''

In [61]:
# результат запроса
df_areas = pd.read_sql_query(query_5_2, connection)
df_areas



C:\Users\asus\AppData\Local\Temp\ipykernel_24208\3394434.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Название города  кол-во работодателей  кол-во вакансий
0                 Россия                   410                0
1              Казахстан                   207                0
2     Московская область                    75                0
3     Краснодарский край                    19                0
4               Беларусь                    18                0
5     Ростовская область                    18                0
6            Азербайджан                    17                0
7  Нижегородская область                    16                0
8   Республика Татарстан                    16                0
9             Узбекистан                    15                0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [62]:
# текст запроса
query_5_3 = f'''SELECT 
    e.name "Название работодателя",
    COUNT(DISTINCT area_id) "Кол-во городов"
FROM employers e 
LEFT JOIN vacancies v ON e.id = v.employer_id
GROUP BY 1
ORDER BY 2 DESC

'''

In [63]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_24208\201667002.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Название работодателя  Кол-во городов
0                      Яндекс             181
1                  Ростелеком             152
2                  Спецремонт             116
3      Поляков Денис Иванович              88
4                   ООО ЕФИН               71
...                       ...             ...
23170                  СДЕЛКА               0
23171            Alandr Group               0
23172                     СДК               0
23173      Lemon Land Lombard               0
23174                  Махина               0

[23175 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [64]:
# текст запроса
query_5_4 = f'''SELECT 
    COUNT(*) as with_no_industry_info
FROM employers e 
LEFT JOIN employers_industries i ON e.id = i.employer_id
LEFT JOIN industries ind ON ind.id = i.industry_id
WHERE ind.name is null 
'''

In [65]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_24208\3493105328.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



with_no_industry_info
0                   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [66]:
# текст запроса
query_5_5 = f'''SELECT 
    e.name,
    Count(*)
FROM employers e 
JOIN employers_industries i ON e.id = i.employer_id
GROUP BY 1

HAVING count(*) = 4
ORDER BY e.name
OFFSET 2
LIMIT 1 
'''

In [67]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_24208\388711819.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  count
0  2ГИС      4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [68]:
# текст запроса
query_5_6 = f'''SELECT 
    COUNT(*) as "Работодатели связанные с разработкой программного обеспечения"
FROM employers e 
JOIN employers_industries i ON e.id = i.employer_id
JOIN industries ind ON ind.id = i.industry_id
WHERE ind.name ilike '%Разработка программного обеспечения' 
'''

In [69]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_24208\821628714.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Работодатели связанные с разработ
0                               3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [70]:
# код для получения списка городов-милионников
import pandas as pd
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
df = pd.read_html(url)[1] #в квадратных скобках индекс таблицы на странице - подбирается путем проб и ошибок

cities = tuple(df['Город']) #SQL принимает кортеж. Выводим данные колонки с городами в виде кортежа


In [71]:
# текст запроса

query_5_7 = f'''(SELECT 
    a.name,
    COUNT(v.name)
FROM employers e
JOIN vacancies v ON v.employer_id = e.id 
JOIN areas a ON v.area_id = a.id
WHERE e.name = 'Яндекс' and a.name in {cities}
GROUP BY 1
ORDER BY 2 DESC) 
UNION ALL
(SELECT 
    'TOTAL',
    COUNT(v.name)
FROM employers e
JOIN vacancies v ON v.employer_id = e.id 
JOIN areas a ON v.area_id = a.id
WHERE e.name = 'Яндекс' and a.name in {cities}
GROUP BY 1)
'''



In [72]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_24208\1965596796.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  count
0             TOTAL    485
1            Москва     54
2   Санкт-Петербург     42
3      Екатеринбург     39
4   Нижний Новгород     36
5       Новосибирск     35
6           Воронеж     32
7         Краснодар     30
8            Самара     26
9               Уфа     26
10           Казань     25
11            Пермь     25
12   Ростов-на-Дону     25
13        Волгоград     24
14       Красноярск     23
15        Челябинск     22
16             Омск     21

***

In [73]:
# выводы по анализу работодателей


#Визулизируем ТОП-10 регинов по кол-ву зарегестрированных работодателей
df_areas = pd.read_sql_query(query_5_2, connection)
df_areas

import plotly.express as px
fig = px.pie(df_areas, values='кол-во работодателей', names='Название города', title='Регионы с наибольшим кол-вом работодателей')
fig.show()


C:\Users\asus\AppData\Local\Temp\ipykernel_24208\606969231.py:5: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



1. Из выборки по запросу видно, что лидером по количеству вакансий является Яндекс. При чем, кол-во вакансий у Яндекса больше почти в 4 раза по сравнению с другими компаниями лидерами на рынке. Среди компаний лидеров по кол-ву вакансий две компании из сектора IT и связи, две - из банковского и одна компания из добывающей сферы промышленности. Можно сделать выводы о самых востребованных сферах экономики и промышленности. 
2. В целом, из данной выборки можно сделать выводы, что основная масса работодателей из базы данных зарегистрированы в России, или субъектах Российской федерации и постсоветском пространстве. Лидером является Россия, Казахстан на втором месте, остальные республики представлены малым кол-вом работодателей. Также есть не значительное количество компаний зарегестрированных в дальнем зарубежье. 
3. Среди компаний лидеров по кол-ву вакансий Яндекс и Ростелеком имеют наиболее обширную территорию размещения вакансий, что говорит о большом кол-ве офисов и представительств этих компаний в стране
4. Треть работодателей, если точнее - 8419, не указывают сферу деятельности. Если проанализировать список вакансий таких работодателей, то окажется что эти работодатели набирают специалистов связанных с IT и разработкой программного обеспечения. В современных компаниях сфера применения IT специалиста очень широка. А компании, специализирующиеся на предоставлении IT услуг часто востребованы в абсолютно разных сферах экономики и промышленности. Видимо, поэтому многие компании при поиске сотрудника не ограничивают себя определенными сферами индустрии. 
 



# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [74]:
# текст запроса
query_6_1 = f'''SELECT 
   COUNT(*) as "Кол-во вакансий в Data Science"
FROM vacancies
WHERE lower(name) like '%data%' or lower(name) like '%данн%'
'''


In [75]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_24208\366702258.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Кол-во вакансий в Data Science
0                            1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [76]:
# текст запроса
query_6_2 = f'''SELECT 
   COUNT(*) as "Вакансий для старта в Data Science"
FROM vacancies
where (experience = 'Нет опыта' or employment = 'Стажировка' or name ilike '%junior%') 
and (name ilike '%data scientist%'
or name ilike '%data science%'
or name ilike '%исследователь данных%'
or name ilike '%machine learning%'
or name ilike '%машинн%обучен%'
or (name not ilike '%HTML%' and name like '%ML%')) 
'''

In [77]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_24208\2267038862.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Вакансий для старта в Data Science
0                                  51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [78]:
# текст запроса
query_6_3 = f'''SELECT 
   COUNT(*) as "Вакансий DS с sql_postgres_skills"
FROM vacancies
where (key_skills ilike '%SQL%' or key_skills ilike '%postgres%' ) 
and (name ilike '%data scientist%'
or name ilike '%data science%'
or name ilike '%исследователь данных%'
or name ilike '%machine learning%'
or name ilike '%machine learning%' 
or name ilike '%машинн%обучен%'
or (name not ilike '%HTML%' and name like '%ML%'))  
'''

In [79]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_24208\3794299702.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Вакансий DS с sql_postgres_skills
0                                201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [80]:
# текст запроса
query_6_4 = f'''SELECT 
   COUNT(*) as "Вакансий DS с python_skill"
FROM vacancies
where key_skills ilike '%PYTHON%' 
and (name ilike '%data scientist%'
or name ilike '%data science%'
or name ilike '%исследователь данных%'
or name ilike '%machine learning%'
or name ilike '%machine learning%' 
or name ilike '%машинн%обучен%'
or (name not ilike '%HTML%' and name like '%ML%'))
'''

In [81]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_24208\4264665179.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Вакансий DS с python_skill
0                         351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [82]:
# текст запроса
query_6_5 = f'''SELECT
    ROUND(AVG(LENGTH(key_skills)-LENGTH(REPLACE(key_skills, CHR(9), ''))+1), 2) cnt_str
FROM vacancies 
WHERE key_skills IS NOT NULL and
(name ilike '%data scientist%'
or name ilike '%data science%'
or name ilike '%исследователь данных%'
or name ilike '%machine learning%'
or name ilike '%машинн%обучен%'
or (name not ilike '%HTML%' and name like '%ML%'))  
'''

In [83]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_24208\3654210517.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



cnt_str
0     6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [84]:
# текст запроса
query_6_6 = f'''SELECT 
experience, 
ROUND(AVG(COALESCE((salary_from + salary_to) / 2, salary_from, salary_to)), 0) AS avg_salary
FROM vacancies 
WHERE (salary_from IS not NULL or salary_to IS not NULL) 
AND (name ilike '%data scientist%'
or name ilike '%data science%'
or name ilike '%исследователь данных%'
or name ilike '%machine learning%'
or name ilike '%машинн%обучен%'
or (name not ilike '%HTML%' and name like '%ML%'))
AND experience = 'От 3 до 6 лет'
GROUP BY 1
'''

In [85]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\asus\AppData\Local\Temp\ipykernel_24208\519194770.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



experience  avg_salary
0  От 3 до 6 лет    243115.0

***

In [86]:
# выводы по предметному анализу

# Посмотрим в каких городах требуются специалисты по DS
query_6_1_1 = f'''SELECT 
    a.name "Город",
   COUNT(*) as "Кол-во вакансий в Data Science"
FROM vacancies v 
JOIN areas a ON v.area_id = a.id
WHERE lower(v.name) like '%data%' or lower(v.name) like '%данн%'
GROUP by 1 
ORDER BY 2 DESC
LIMIT 5'''
ds_ciiies = pd.read_sql_query(query_6_1_1, connection)
display(ds_ciiies)

# Выведем города с вакансиями для новичков
query_6_2_1 = f'''SELECT 
   a.name "Город",
   count(*) "Вакансий для начинающего DS"
FROM vacancies v 
JOIN areas a ON v.area_id = a.id
where (experience = 'Нет опыта' or employment = 'Стажировка' or v.name ilike '%junior%') 
and (v.name ilike '%data scientist%'
or v.name ilike '%data science%'
or v.name ilike '%исследователь данных%'
or v.name ilike '%machine learning%'
or v.name ilike '%машинн%обучен%'

or (v.name not ilike '%HTML%' and v.name like '%ML%'))
GROUP BY 1
ORDER BY 2 DESC
LIMIT 5'''
for_start = pd.read_sql_query(query_6_2_1, connection)
display(for_start)


C:\Users\asus\AppData\Local\Temp\ipykernel_24208\963393083.py:13: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Город  Кол-во вакансий в Data Science
0           Москва                             619
1  Санкт-Петербург                             164
2           Алматы                             105
3            Минск                              89
4      Новосибирск                              57

C:\Users\asus\AppData\Local\Temp\ipykernel_24208\963393083.py:33: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Город  Вакансий для начинающего DS
0           Москва                           19
1  Санкт-Петербург                            9
2      Новосибирск                            5
3     Екатеринбург                            2
4            Минск                            2

1. В нашей базе данных имеется 1771 вакансия для специалиста Data Science (далее по тексту DS). Причём на Москву приходится треть всех предложений. В пятерку городов с наибольшим кол-вом предложений для DS специалиста входят также столица Казахстана и Белоруси. Вакансии ожидаемо распределены по количеству населения в городах.
2. Для начинающего специалиста имеется 51 предложение, что составляет около 3% от всех предложений на рынке труда. Если сравнить два запроса query_6_1_1 и query_6_2_1 из выводов по предметному анализу, то увидим, что в столице Казахстана новички не нужны. Туда приглашают на работу состоявшихся специалистов. Если сравнить распределение вакансий из двух выборок, то окажется, что лидером вакансий для новичков является Новосибирск - в этом городе около 10% вакансий для DS приходится на новичков; на втором месте Санкт-Петербург, потом - Москва.
3. Менее чем в 10% вакансий для DS указана необходимость знания SQL Postgres.
4. Менее чем в 20% работодатели при поиске специалиста по DS указывают необходимость знания Python.
5. Из запроса 6_5 видно, что работодатели указывают около 6 ключевых навыков, которые должен иметь специалист DS 
6. Средняя зарплата на вакансию DS для специалиста с опытом работы от 3 до 6 лет составляет 243115 рублей


# Общий вывод по проекту

In [87]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

#Пересчитаем всех работодателей с открытыми вакансиями
query_7_1 = '''SELECT *
FROM
(SELECT 
    COUNT(distinct e.name) "Работодателей с вакансиями"
FROM employers e
JOIN vacancies v ON v.employer_id = e.id) as employers,

(SELECT 
    COUNT(id) "Всего работодателей"
FROM employers) as total'''
real_jobs = pd.read_sql_query(query_7_1, connection)
display(real_jobs)

#Посчитаем сколько из всех зарегистрированных работодателей предлагают работу в городах-миллиониках
query_7_2 = f'''SELECT 
    COUNT(distinct e.id) "Реальных работодателей в городах"
FROM employers e
JOIN vacancies v ON v.employer_id = e.id 
JOIN areas a ON v.area_id = a.id
WHERE a.name in {cities}'''
jobs_in_cities= pd.read_sql_query(query_7_2, connection)
display(jobs_in_cities)

# посмотрим сколько вакансий от компаний лидеров приходится на города миллионики и сравним с общим кол-вом вакансий от этих компаний
query_5_7_1 = f'''SELECT *
from 
(SELECT 
    e.name,
    COUNT(*) "Всего вакансий"
FROM vacancies v 
JOIN employers e ON v.employer_id = e.id
GROUP BY 1 
ORDER BY 2 DESC
LIMIT 5) as total,

(SELECT 
    COUNT(*) "Вакансий в городах миллиониках",
    e.name
FROM employers e
JOIN vacancies v ON v.employer_id = e.id 
JOIN areas a ON v.area_id = a.id
WHERE e.name in ('Яндекс', 'Ростелеком', 'Тинькофф', 'СБЕР', 'Газпром нефть') and a.name in {cities}
GROUP BY 2 
ORDER bY 1 DESC
LIMIT 5) as million
WHERE total.name = million.name'''
job_numbers = pd.read_sql_query(query_5_7_1, connection)
display(job_numbers)

#В запросе 5_8 - 5_10 выведем компании связанные с разработкой и обслуживанием программного обеспечения
#Проанализируем какие компании не указывают сферы своей деятельности при наборе сотрудников
query_5_8 = f'''SELECT 
    e.name "Название компании",
    v.name "Вакансия"
FROM employers e 
LEFT JOIN vacancies v on v.employer_id = e.id
LEFT JOIN employers_industries i ON e.id = i.employer_id
LEFT JOIN industries ind ON ind.id = i.industry_id
WHERE ind.name is null'''
ind_null = pd.read_sql_query(query_5_8, connection)
display(ind_null.head(20))

#Проанализируем какие компании указывают 4 и более сфер деятельности и при этом отфильтруем эти компании по условию, что они не связаны с разработкой программного обеспечения 
query_5_9 = f'''SELECT 
    DISTINCT e.name "Название работодателя",
    v.name "Вакансия",
    ind.name "Индустрия",
    Count(*) "Подсчет сфер деятельности"
FROM employers e 
JOIN employers_industries i ON e.id = i.employer_id
JOIN industries ind ON ind.id = i.industry_id
JOIN vacancies v ON v.employer_id = e.id
WHERE ind.name != 'Разработка программного обеспечения'
GROUP BY 1, 2, 3
HAVING count(*) >= 4'''
check_employers = pd.read_sql_query(query_5_9, connection)
display(check_employers.head(20))

#Посчитаем количество таких компаний, у которых указано много сфер деятельности, но не указана сфера "Разработка программного обеспечения"
query_5_10 = f'''with ind_count as (SELECT 
    e.name "Название работодателя",
    v.name "Вакансия",
    ind.name "Индустрия",
    Count(*) "Подсчет сфер деятельности"
FROM employers e 
JOIN employers_industries i ON e.id = i.employer_id
JOIN industries ind ON ind.id = i.industry_id
JOIN vacancies v ON v.employer_id = e.id
WHERE ind.name != 'Разработка программного обеспечения'
GROUP BY 1, 2, 3
HAVING count(*) >= 4)

SELECT 
   COUNT(*) "Компаний с большим числом сфер деятельности"
FROM ind_count'''
total_companies = pd.read_sql_query(query_5_10, connection)
display(total_companies)

#Пересчитаем все вакансии для DS с опытом от 3 до 6 лет
query_7_2 = f'''SELECT 
count(*) "Всех вакансий для DS"
FROM vacancies 
WHERE  
name ilike '%data scientist%'
or name ilike '%data science%'
or name ilike '%исследователь данных%'
or name ilike '%machine learning%'
or name ilike '%машинн%обучен%'
or (name not ilike '%HTML%' and name like '%ML%')
AND experience = 'От 3 до 6 лет' '''
total_DS = pd.read_sql_query(query_7_2, connection)
display(total_DS)

# Пересчитаем вакансии с указанным одним из параметров зарплаты
query_7_3 = f'''SELECT 
    COUNT(*) "Вакансий с указанным одним из параметров"
FROM vacancies 
WHERE (salary_from IS not NULL or salary_to IS not NULL) 
AND (name ilike '%data scientist%'
or name ilike '%data science%'
or name ilike '%исследователь данных%'
or name ilike '%machine learning%'
or name ilike '%машинн%обучен%'
or (name not ilike '%HTML%' and name like '%ML%'))
AND experience = 'От 3 до 6 лет' '''
DS_with_salary = pd.read_sql_query(query_7_3, connection)
display(DS_with_salary)


C:\Users\asus\AppData\Local\Temp\ipykernel_24208\2383848574.py:15: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Работодателей с вакансиями  Всего работодателей
0                       14766                23501

C:\Users\asus\AppData\Local\Temp\ipykernel_24208\2383848574.py:25: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Реальных работодателей в городах
0                              7621

C:\Users\asus\AppData\Local\Temp\ipykernel_24208\2383848574.py:51: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  Всего вакансий  Вакансий в городах миллиониках  \
0         Яндекс            1933                             485   
1     Ростелеком             491                             214   
2       Тинькофф             444                             285   
3           СБЕР             428                             364   
4  Газпром нефть             331                             245   

            name  
0         Яндекс  
1     Ростелеком  
2       Тинькофф  
3           СБЕР  
4  Газпром нефть

C:\Users\asus\AppData\Local\Temp\ipykernel_24208\2383848574.py:64: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Название компании  \
0                                          Спецремонт   
1                                      Диалог Регионы   
2                                        ООО ЮрСпектр   
3                                            АО ЛЕПСЕ   
4                                           ООО ЕФИН    
5          Филиал ФГБУ ЦЖКУ Минобороны России по РВСН   
6                                            Интелком   
7                                     Софт Инжиниринг   
8                                  Страна Девелопмент   
9     ФГБУ Федеральный центр охраны здоровья животных   
10                                               Эр-1   
11                                            Риадент   
12    ФГБУ Федеральный центр охраны здоровья животных   
13                                             Диплей   
14  Филиал ФКУ Налог-Сервис ФНС России в Новосибир...   
15  Филиал ФКУ Налог-Сервис ФНС России в Новосибир...   
16  Новосибирский отдел вневедомственной охраны - ...   
17                                  Технология-Тюмень   
18                                         ТехноГород   
19  Новосибирский отдел вневедомственной охраны - ...   

                                             Вакансия  
0                                 Компьютерный Мастер  
1                             Системный администратор  
2                                           Корректор  
3                                 Инженер-программист  
4                              Ведущий программист 1C  
5                               Инженер автохозяйства  
6      Ночной Специалист службы технической поддержки  
7            Разработчик видеоигр (Unity разработчик)  
8                     Разработчик BackEnd (Team Lead)  
9                            Программист Java, Spring  
10                          CPO/Chief Product Officer  
11                        Web-разработчик/программист  
12               Системный аналитик / Бизнес-аналитик  
13                               Технический писатель  
14    Ведущий специалист отдела технической поддержки  
15                                      ИТ-специалист  
16                                 Главный специалист  
17                             Технический специалист  
18  Инженер по разработке проектов организации дор...  
19                                 Главный специалист

C:\Users\asus\AppData\Local\Temp\ipykernel_24208\2383848574.py:80: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Название работодателя                                           Вакансия  \
0                1C-Рарус                           PHP-разработчик (middle)   
1                1C-Рарус                                     Программист 1С   
2                1C-Рарус  Специалист технической поддержки облачных серв...   
3   1С:Франчайзинг Ваниев                            Программист 1С (Middle)   
4   1С:Франчайзинг Ваниев                            Программист 1С (Middle)   
5                    2ГИС                   Frontend-разработчик в Web-карту   
6                    2ГИС                   Frontend-разработчик в Web-карту   
7                    2ГИС                   Frontend-разработчик в Web-карту   
8                    2ГИС                   Frontend-разработчик в Web-карту   
9                    2ГИС              Frontend-разработчик в команду Online   
10                   2ГИС              Frontend-разработчик в команду Online   
11                   2ГИС              Frontend-разработчик в команду Online   
12                   2ГИС              Frontend-разработчик в команду Online   
13                   2ГИС            Kotlin/Java-разработчик в команду World   
14                   2ГИС            Kotlin/Java-разработчик в команду World   
15                   2ГИС            Kotlin/Java-разработчик в команду World   
16                   2ГИС            Kotlin/Java-разработчик в команду World   
17                   2ГИС  Middle/Senior Python-разработчик в команду Доб...   
18                   2ГИС  Middle/Senior Python-разработчик в команду Доб...   
19                   2ГИС  Middle/Senior Python-разработчик в команду Доб...   

                                            Индустрия  \
0   Системная интеграция,  автоматизации технологи...   
1   Системная интеграция,  автоматизации технологи...   
2   Системная интеграция,  автоматизации технологи...   
3                               Консалтинговые услуги   
4   Системная интеграция,  автоматизации технологи...   
5   Интернет-компания (поисковики, платежные систе...   
6                   Колл-центры, диспетчерские службы   
7                               Консалтинговые услуги   
8   Производство мультимедиа, контента, редакторск...   
9   Интернет-компания (поисковики, платежные систе...   
10                  Колл-центры, диспетчерские службы   
11                              Консалтинговые услуги   
12  Производство мультимедиа, контента, редакторск...   
13  Интернет-компания (поисковики, платежные систе...   
14                  Колл-центры, диспетчерские службы   
15                              Консалтинговые услуги   
16  Производство мультимедиа, контента, редакторск...   
17  Интернет-компания (поисковики, платежные систе...   
18                  Колл-центры, диспетчерские службы   
19                              Консалтинговые услуги   

    Подсчет сфер деятельности  
0                           4  
1                           6  
2                           8  
3                           5  
4                           5  
5                           6  
6                           6  
7                           6  
8                           6  
9                           4  
10                          4  
11                          4  
12                          4  
13                          4  
14                          4  
15                          4  
16                          4  
17                          4  
18                          4  
19                          4

C:\Users\asus\AppData\Local\Temp\ipykernel_24208\2383848574.py:100: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Компаний с большим числом сфер дея
0                                1725

C:\Users\asus\AppData\Local\Temp\ipykernel_24208\2383848574.py:115: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Всех вакансий для DS
0                   424

C:\Users\asus\AppData\Local\Temp\ipykernel_24208\2383848574.py:130: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Вакансий с указанным одним из пара
0                                  27

Обобщим выводы:
1. Всего в базе данных 49147 вакансий - 23856 вакансий находятся в городах миллиониках. В нашей базе данных зарегестрированы 23501 работодателей и только 14766 имеют открытые вакансии. При помощи запроса query_7_2 мы обнаружим что половина всех компаний, которые предлагают работу также находятся в городах миллиониках. Очевидно, что чем меньше город, тем компаний в нем меньше.
Можно сделать вывод, что каждая компания в среднем разместила 3 вакансии. 
2. Поскольку о зарплате есть информация только от половины всех объявлений, показатель средней зарплаты в 90796 рублей (средний показатель верхнего порога зарплаты + средний показатель нижнего порога зарплаты)/2 не является объективным. В большинстве случаев работодатели не указывают размер зарплаты, если зарплата низкая. Поэтому можно сделать предположение, что средняя зарплата окажется процентов на 25 ниже.
3. При помощи запроса query_5_7_1 ещё раз убедимся, что у компаний лидеров по количеству вакансий также около половины приходятся на города миллионики.
4. При помощи запросов 5_8 - 5_10 показано количество компаний, которые указывают 4 и более сфер деятельности. А также можно убедится в том, что эти компании так или иначе связаны с разработкой и обслуживанием программного обеспечения. Если суммировать компании в query_5_4, которые не указывают сферу деятельности и компании, которые рассматриваются при помощи запросов 5_6 - 5_10, то мы получим общее количество компаний, которым требуются различные IT специалисты. Таких компаний в нашей базе данных - 13697, при том что компаний с реальными открытыми вакансиями 14766. Получается, что из всех компаний, разместивших вакансии в нашей базе данных, 92% ищут как минимум 1 специалиста в сфере разработки программного обеспечения. 

5. Если провести дополнительный анализ вакансий DS для специалиста с опытом работы от 3 до 6 лет, то окажется что подсчет средней зарплаты не объективен. Достаточно пересчитать общее кол-во вакансий для DS и посчитать кол-во тех вакансий, в которых указан хоть один из параметров зарплаты, на основании которых мы высчитывали решение задания 6.6. 